### PR TITLE
IGNITE-20423 Make org.apache.ignite.internal.manager.Producer an interface

### PR DESCRIPTION
**Pull request by tkalkirill**
https://issues.apache.org/jira/browse/IGNITE-20423

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/event/EventListener.java [hunk 1]
```diff
@@ -15,35 +15,22 @@
...
 
 import java.util.concurrent.CompletableFuture;
 import org.jetbrains.annotations.Nullable;
 
-/**
- * The listener handles events from a producer.
- *
- * @see Producer#listen(Event, EventListener)
- */
+/** The listener handles events from a {@link EventProducer event producer}. */
```
Comment (sashapolo) [inline modules/core/src/main/java/org/apache/ignite/internal/event/EventListener.java]:
```suggestion
/** The listener handles events from an {@link EventProducer event producer}. */
```

File: modules/core/src/main/java/org/apache/ignite/internal/event/EventParameters.java [hunk 2]
```diff
@@ -15,15 +15,10 @@
...
 
-package org.apache.ignite.internal.manager;
+package org.apache.ignite.internal.event;
 
-/**
- * Event parameters. This type passed to the event listener.
- *
- * @see Producer#fireEvent(Event, EventParameters, Throwable)
- */
+/** Event parameters. This type passed to the {@link EventListener#notify(EventParameters, Throwable)}. */
```
Comment (sashapolo) [inline modules/core/src/main/java/org/apache/ignite/internal/event/EventParameters.java:20]:
```suggestion
/** Event parameters. This type is passed to the {@link EventListener#notify(EventParameters, Throwable)}. */
```

File: modules/core/src/main/java/org/apache/ignite/internal/event/AbstractEventProducer.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+ *
+ * <p>Allows to {@link #listen add} and {@link #removeListener remove} event listeners for events, as well as
+ * {@link #fireEvent fire events}.</p>
+ */
+public abstract class AbstractEventProducer<T extends Event, P extends EventParameters> implements EventProducer<T, P> {
+    private final ConcurrentHashMap<T, List<EventListener<P>>> listenersByEvent = new ConcurrentHashMap<>();
+
+    @Override
+    public void listen(T evt, EventListener<? extends P> listener) {
+        listenersByEvent.computeIfAbsent(evt, evtKey -> new CopyOnWriteArrayList<>()).add((EventListener<P>) listener);
```
Comment (sashapolo) [inline modules/core/src/main/java/org/apache/ignite/internal/event/AbstractEventProducer.java:41]: Can we use a manual copy approach here instead of using `CopyOnWriteArrayList`? We are always accessing it inside of `compute` anyway

File: modules/core/src/main/java/org/apache/ignite/internal/event/AbstractEventProducer.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    protected CompletableFuture<Void> fireEvent(T evt, P params, @Nullable Throwable err) {
+        List<EventListener<P>> listeners = listenersByEvent.get(evt);
+
+        if (listeners == null) {
+            return completedFuture(null);
+        }
+
+        List<CompletableFuture<?>> futures = new ArrayList<>();
+
+        for (EventListener<P> listener : listeners) {
```
Comment (sashapolo) [inline modules/core/src/main/java/org/apache/ignite/internal/event/AbstractEventProducer.java:70]: Since `listeners` is a `List`, we can declare `futures` as an array and remove the copy below.

File: modules/core/src/main/java/org/apache/ignite/internal/event/Event.java [hunk 5]
```diff
@@ -15,12 +15,8 @@
...
 
-package org.apache.ignite.internal.manager;
+package org.apache.ignite.internal.event;
 
-/**
- * The event cas which is produced by event producer component.
- *
- * @see Producer#fireEvent(Event, EventParameters, Throwable)
- */
+/** The event which is produced by a component. */
```
Comment (sashapolo) [inline modules/core/src/main/java/org/apache/ignite/internal/event/Event.java:20]:
```suggestion
/** An event which is produced by a component. */
```

File: modules/core/src/main/java/org/apache/ignite/internal/event/EventListener.java [hunk 6]
```diff
@@ -15,35 +15,22 @@
...
 
 import java.util.concurrent.CompletableFuture;
 import org.jetbrains.annotations.Nullable;
 
-/**
- * The listener handles events from a producer.
- *
- * @see Producer#listen(Event, EventListener)
- */
+/** The listener handles events from a event producer. */
```
Comment (sashapolo) [inline modules/core/src/main/java/org/apache/ignite/internal/event/EventListener.java:23]:
```suggestion
/** A listener that handles events from an event producer. */
```

File: modules/core/src/main/java/org/apache/ignite/internal/event/EventListener.java [hunk 7]
```diff
@@ -15,35 +15,22 @@
...
 @FunctionalInterface
 public interface EventListener<P extends EventParameters> {
     /**
      * Notifies the listener about an event.
      *
-     * @param parameters Parameters provide a properties of the event. This attribute cannot be {@code null}.
+     * @param parameters Parameters provide a properties of the event.
```
Comment (sashapolo) [inline modules/core/src/main/java/org/apache/ignite/internal/event/EventListener.java:29]:
```suggestion
     * @param parameters Properties of the event.
```

File: modules/core/src/main/java/org/apache/ignite/internal/event/EventListener.java [hunk 8]
```diff
@@ -15,35 +15,22 @@
...
      * @param exception  Exception which is happened during the event produced or {@code null}.
```
Comment (sashapolo) [inline modules/core/src/main/java/org/apache/ignite/internal/event/EventListener.java:30]:
```suggestion
     * @param exception  Exception which has happened when producing the event or {@code null}.
```

File: modules/core/src/main/java/org/apache/ignite/internal/event/EventListener.java [hunk 9]
```diff
@@ -15,35 +15,22 @@
...
      * @return Completable future, which is completed when event handling is finished. The {@code True} value of the future
```
Comment (sashapolo) [inline modules/core/src/main/java/org/apache/ignite/internal/event/EventListener.java:31]:
```suggestion
     * @return Completable future, which is completed when event handling is finished. The {@code true} value of the future
```

File: modules/core/src/main/java/org/apache/ignite/internal/event/EventListener.java [hunk 10]
```diff
@@ -15,35 +15,22 @@
...
      *     means that the event is handled and a listener will be removed, {@code false} is the listener will stay listen.
```
Comment (sashapolo) [inline modules/core/src/main/java/org/apache/ignite/internal/event/EventListener.java:32]:
```suggestion
     *     means that the event has been handled and a listener will be removed, {@code false} is that the listener will continue listening.
```

File: modules/core/src/main/java/org/apache/ignite/internal/event/EventListener.java [hunk 11]
```diff
@@ -15,35 +15,22 @@
...
      *     This future must not be completed with {@code null} value.
```
Comment (sashapolo) [inline modules/core/src/main/java/org/apache/ignite/internal/event/EventListener.java:33]:
```suggestion
     *     This future will never be completed with {@code null} value.
```

File: modules/core/src/main/java/org/apache/ignite/internal/event/EventProducer.java [hunk 12]
```diff
@@ -15,26 +15,24 @@
...
-package org.apache.ignite.internal.manager;
+package org.apache.ignite.internal.event;
 
-import org.apache.ignite.lang.IgniteInternalCheckedException;
-
-/**
- * The exception notifies a listener when the listener was removed from queue and never receive a notification again.
- */
-public class ListenerRemovedException extends IgniteInternalCheckedException {
+/** Allows to {@link #listen add} and {@link #removeListener remove} event listeners that the component will fire. */
```
Comment (sashapolo) [inline modules/core/src/main/java/org/apache/ignite/internal/event/EventProducer.java:20]:
```suggestion
/** Allows to {@link #listen add} and {@link #removeListener remove} event listeners that the component will produce. */
```

File: modules/core/src/main/java/org/apache/ignite/internal/event/EventProducer.java [hunk 13]
```diff
@@ -15,26 +15,24 @@
...
+public interface EventProducer<T extends Event, P extends EventParameters> {
     /**
-     * Default constructor.
+     * Registers an event listener. When the event predicate returns true it would never invoke after, otherwise this predicate would
+     * receive an event again.
```
Comment (sashapolo) [inline modules/core/src/main/java/org/apache/ignite/internal/event/EventProducer.java:24]:
```suggestion
     * Registers an event listener. If {@link EventListener#notify} returns {@code true}, it would never be invoked again.
```

File: modules/core/src/test/java/org/apache/ignite/internal/event/EventProducerTest.java [hunk 14]
```diff
@@ -45,26 +42,23 @@ public void simpleAsyncTest() {
 
         CompletableFuture<?> eventHandleFuture = producer.fireEvent(TestEvent.TEST, new TestEventParameters(0L));
 
-        assertFalse(eventHandleFuture.isDone());
+        assertThat(eventHandleFuture, willTimeoutFast());
```
Comment (sashapolo) [inline modules/core/src/test/java/org/apache/ignite/internal/event/EventProducerTest.java:45]: Why did you change this check?

File: modules/core/src/test/java/org/apache/ignite/internal/event/EventProducerTest.java [hunk 15]
```diff
@@ -45,26 +42,23 @@ public void simpleAsyncTest() {
...
 
         future.complete(true);
 
-        assertTrue(eventHandleFuture.isDone());
+        assertThat(eventHandleFuture, willCompleteSuccessfully());
```
Comment (sashapolo) [inline modules/core/src/test/java/org/apache/ignite/internal/event/EventProducerTest.java:49]: And this?

File: modules/core/src/test/java/org/apache/ignite/internal/event/EventProducerTest.java [hunk 16]
```diff
@@ -74,31 +68,28 @@ public void stopListenTest() {
...
+                        toRemoveFuture.complete(null);
 
-                            return completedFuture(false);
-                        },
-                        t -> {}
-                    )
-                    : createEventListener((p, e) -> completedFuture(false), t -> {});
+                        return completedFuture(false);
+                    }
+            )
```
Comment (sashapolo) [inline modules/core/src/test/java/org/apache/ignite/internal/event/EventProducerTest.java:91]: This formatting looks wrong

File: modules/core/src/main/java/org/apache/ignite/internal/event/EventListener.java (hunk 1 above)
Comment (tkalkirill) [reply modules/core/src/main/java/org/apache/ignite/internal/event/EventListener.java]: fix it

File: modules/core/src/main/java/org/apache/ignite/internal/event/EventParameters.java (hunk 2 above)
Comment (tkalkirill) [reply modules/core/src/main/java/org/apache/ignite/internal/event/EventParameters.java:20]: fix it

File: modules/core/src/main/java/org/apache/ignite/internal/event/Event.java (hunk 5 above)
Comment (tkalkirill) [reply modules/core/src/main/java/org/apache/ignite/internal/event/Event.java:20]: fix it

File: modules/core/src/main/java/org/apache/ignite/internal/event/EventListener.java (hunk 6 above)
Comment (tkalkirill) [reply modules/core/src/main/java/org/apache/ignite/internal/event/EventListener.java:23]: fix it

File: modules/core/src/main/java/org/apache/ignite/internal/event/EventListener.java (hunk 7 above)
Comment (tkalkirill) [reply modules/core/src/main/java/org/apache/ignite/internal/event/EventListener.java:29]: fix it

File: modules/core/src/main/java/org/apache/ignite/internal/event/EventListener.java (hunk 8 above)
Comment (tkalkirill) [reply modules/core/src/main/java/org/apache/ignite/internal/event/EventListener.java:30]: fix it

File: modules/core/src/main/java/org/apache/ignite/internal/event/EventListener.java (hunk 9 above)
Comment (tkalkirill) [reply modules/core/src/main/java/org/apache/ignite/internal/event/EventListener.java:31]: fix it

File: modules/core/src/main/java/org/apache/ignite/internal/event/EventListener.java (hunk 10 above)
Comment (tkalkirill) [reply modules/core/src/main/java/org/apache/ignite/internal/event/EventListener.java:32]: fix it

File: modules/core/src/main/java/org/apache/ignite/internal/event/EventListener.java (hunk 11 above)
Comment (tkalkirill) [reply modules/core/src/main/java/org/apache/ignite/internal/event/EventListener.java:33]: fix it

File: modules/core/src/main/java/org/apache/ignite/internal/event/EventProducer.java (hunk 12 above)
Comment (tkalkirill) [reply modules/core/src/main/java/org/apache/ignite/internal/event/EventProducer.java:20]: fix it

File: modules/core/src/main/java/org/apache/ignite/internal/event/EventProducer.java (hunk 13 above)
Comment (tkalkirill) [reply modules/core/src/main/java/org/apache/ignite/internal/event/EventProducer.java:24]: fix it

File: modules/core/src/test/java/org/apache/ignite/internal/event/EventProducerTest.java (hunk 14 above)
Comment (tkalkirill) [reply modules/core/src/test/java/org/apache/ignite/internal/event/EventProducerTest.java:45]: Return it

File: modules/core/src/test/java/org/apache/ignite/internal/event/EventProducerTest.java (hunk 15 above)
Comment (tkalkirill) [reply modules/core/src/test/java/org/apache/ignite/internal/event/EventProducerTest.java:49]: This way we will check that the future completed successfully without errors instead of two checks.

File: modules/core/src/test/java/org/apache/ignite/internal/event/EventProducerTest.java (hunk 16 above)
Comment (tkalkirill) [reply modules/core/src/test/java/org/apache/ignite/internal/event/EventProducerTest.java:91]: fix it

File: modules/core/src/test/java/org/apache/ignite/internal/event/EventProducerTest.java (hunk 15 above)
Comment (sashapolo) [reply modules/core/src/test/java/org/apache/ignite/internal/event/EventProducerTest.java:49]: I think the idea was that the future would complete immediately

File: modules/core/src/main/java/org/apache/ignite/internal/event/AbstractEventProducer.java (hunk 3 above)
Comment (tkalkirill) [reply modules/core/src/main/java/org/apache/ignite/internal/event/AbstractEventProducer.java]: Did it

File: modules/core/src/main/java/org/apache/ignite/internal/event/AbstractEventProducer.java (hunk 4 above)
Comment (tkalkirill) [reply modules/core/src/main/java/org/apache/ignite/internal/event/AbstractEventProducer.java]: Fix it

File: modules/core/src/main/java/org/apache/ignite/internal/event/AbstractEventProducer.java [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+public abstract class AbstractEventProducer<T extends Event, P extends EventParameters> implements EventProducer<T, P> {
+    private final ConcurrentHashMap<T, List<EventListener<P>>> listenersByEvent = new ConcurrentHashMap<>();
+
+    @Override
+    public void listen(T evt, EventListener<? extends P> listener) {
+        listenersByEvent.compute(evt, (evt0, listeners) -> {
+            List<EventListener<P>> newListeners;
+
+            if (listeners == null) {
+                newListeners = new ArrayList<>(1);
```
Comment (sashapolo) [inline modules/core/src/main/java/org/apache/ignite/internal/event/AbstractEventProducer.java:45]: The idea was that you would use `List.of` and `List.copyOf`, but this approach is also ok, I guess

File: modules/core/src/test/java/org/apache/ignite/internal/event/EventProducerTest.java (hunk 15 above)
Comment (tkalkirill) [reply modules/core/src/test/java/org/apache/ignite/internal/event/EventProducerTest.java:48]: fix it

**Review by sashapolo — APPROVED** [review]
